Firestore/core/src/firebase/firestore/util/executor.h

```diff
@@ -0,0 +1,128 @@
+  void Cancel() {
+    cancel_func_();
+  }
+
+  // Internal use only.
+  explicit DelayedOperation(std::function<void()>&& cancel_func)
+      : cancel_func_{std::move(cancel_func)} {
+  }
+
+ private:
+  std::function<void()> cancel_func_;
+};
+
+namespace internal {
+
+// An interface to a platform-specific executor of asynchronous tasks
+// ("operations").
+//
+// Operations may be scheduled for immediate or delayed execution. Operations
+// scheduled for the exact same time are FIFO ordered.
+//
+// The operations are executed sequentially; only a single operation is executed
+// at any given time.
+//
+// Delayed operations may be canceled if they have not already been run.
+class Executor {
+ public:
+  using Tag = int;
+  using Operation = std::function<void()>;
+  using Milliseconds = std::chrono::milliseconds;
+
+  // Operations scheduled for future execution are tagged to allow retreiving
+  // the later. The tag is entirely opaque for the executor; in particular,
+  // uniqueness of tags is not enforced.
+  struct TaggedOperation {
+    TaggedOperation() {
+    }
+    TaggedOperation(const Tag tag, Operation&& operation)
+        : tag{tag}, operation{std::move(operation)} {
+    }
+    Tag tag = 0;
+    Operation operation;
+  };
+
+  virtual ~Executor() {
+  }
+
+  // Schedules the `operation` to be asynchronously executed as soon as
+  // possible. If called in quick succession, the operations will be
+  // FIFO-ordered.
+  virtual void Execute(Operation&& operation) = 0;
+  // Like `Execute`, but blocks until the `operation` finishes, consequently
+  // draining immediate operations from the executor.
+  virtual void ExecuteBlocking(Operation&& operation) = 0;
+  // Scheduled the given `operation` to be executed after `delay` milliseconds
+  // from now, and returns a handle that allows to cancel the operation
+  // (provided it hasn't been run already). The operation is tagged to allow
+  // retrieving it later.
+  //
+  // `delay` must be non-negative; use `Execute` to schedule operations for
+  // immediate execution.
+  virtual DelayedOperation ScheduleExecution(Milliseconds delay,
+                                             TaggedOperation&& operation) = 0;
+
+  // Checks for the caller whether it is being invoked by this executor.
+  virtual bool IsAsyncCall() const = 0;
+  // Returns some sort of an identifier for the current execution context. The
+  // only guarantee is that it will return different values depending on whether
+  // this function is invoked by this executor or not.
+  virtual std::string GetInvokerId() const = 0;
+
+  // Checks whether an operation tagged with the given `tag` is currently
+  // scheduled for future execution.
+  virtual bool IsScheduled(Tag tag) const = 0;
+  // Checks whether there are any scheduled operations pending execution.
+  // Operations scheduled for immediate execution don't count, even if they
+  // haven't been run already.
+  virtual bool IsScheduleEmpty() const = 0;
```

Comment:
This kind of method makes me nervous because they frequently get used in ways that create races. For example:

```c++
if (IsScheduleEmpty()) {
  Schedule([] {
    // something that depends on the schedule being empty
  });
}
```
Better to expose the composite action directly in this interface or abstain from exposing this at all.

Reply:
(also tagging @rsgowman Rich, who made a similar comment below)

I'll address it better tomorrow, but wanted to write a quick reply. As it stands, I think this is okay, because this function is only ever called from inside `ExecuteBlocking`. However, it's a valid point that this function is prone to misuse. It basically exists to make `AsyncQueue::RunScheduledOperationsUntil` possible -- I tried to make it translate into more-or-less meaningful operations on the `Executor`.

Reply:
Update: I changed `PopFromSchedule` to return an `optional` and removed `IsScheduleEmpty`.